sideload.go

```diff
@@ -0,0 +1,240 @@
+package kapacitor
+
+import (
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"text/template"
+	text "text/template"
+
+	"github.com/influxdata/kapacitor/bufpool"
+	"github.com/influxdata/kapacitor/edge"
+	"github.com/influxdata/kapacitor/keyvalue"
+	"github.com/influxdata/kapacitor/models"
+	"github.com/influxdata/kapacitor/pipeline"
+	"github.com/influxdata/kapacitor/services/sideload"
+	"github.com/pkg/errors"
+)
+
+type SideloadNode struct {
+	node
+	s          *pipeline.SideloadNode
+	source     sideload.Source
+	orderTmpls []orderTmpl
+
+	order []string
+
+	bufferPool *bufpool.Pool
+}
+
+// Create a new  SideloadNode which shifts points and batches in time.
+func newSideloadNode(et *ExecutingTask, n *pipeline.SideloadNode, d NodeDiagnostic) (*SideloadNode, error) {
+	sn := &SideloadNode{
+		node:       node{Node: n, et: et, diag: d},
+		s:          n,
+		bufferPool: bufpool.New(),
+		order:      make([]string, len(n.OrderList)),
+		orderTmpls: make([]orderTmpl, len(n.OrderList)),
+	}
+	u, err := url.Parse(n.Source)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme != "file" {
+		return nil, fmt.Errorf("unsupported source scheme %q, must be 'file'", u.Scheme)
+	}
+	if !filepath.IsAbs(u.Path) {
+		return nil, fmt.Errorf("sideload source path must be absolute %q", u.Path)
+	}
+	src, err := et.tm.SideloadService.Source(u.Path)
+	if err != nil {
+		return nil, err
+	}
+	sn.source = src
+	for i, o := range n.OrderList {
+		op, err := newOrderTmpl(o, sn.bufferPool)
+		if err != nil {
+			return nil, err
+		}
+		sn.orderTmpls[i] = op
+	}
+	sn.node.runF = sn.runSideload
+	sn.node.stopF = sn.stopSideload
+	return sn, nil
+}
+
+func (n *SideloadNode) runSideload([]byte) error {
+	consumer := edge.NewConsumerWithReceiver(
+		n.ins[0],
+		edge.NewReceiverFromForwardReceiverWithStats(
+			n.outs,
+			edge.NewTimedForwardReceiver(n.timer, n),
+		),
+	)
+	return consumer.Consume()
+}
+func (n *SideloadNode) stopSideload() {
+	n.source.Close()
+}
+
+type orderTmpl struct {
+	raw        string
+	tmpl       *text.Template
+	bufferPool *bufpool.Pool
+}
+
+func newOrderTmpl(text string, bp *bufpool.Pool) (orderTmpl, error) {
+	t, err := template.New("order").Parse(text)
+	if err != nil {
+		return orderTmpl{}, err
+	}
+	return orderTmpl{
+		raw:        text,
+		tmpl:       t,
+		bufferPool: bp,
+	}, nil
+}
+
+func (t orderTmpl) Path(tags models.Tags) (string, error) {
+	buf := t.bufferPool.Get()
+	defer t.bufferPool.Put(buf)
+	err := t.tmpl.Execute(buf, tags)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func (n *SideloadNode) doSideload(p edge.FieldsTagsTimeSetter) {
+	for i, o := range n.orderTmpls {
+		p, err := o.Path(p.Tags())
+		if err != nil {
+			n.diag.Error("failed to evaluate order template", err, keyvalue.KV("order", o.raw))
+			return
+		}
+		n.order[i] = p
+	}
+	if len(n.s.Fields) > 0 {
```

Comment:
Sucks that the logic for fields and tags can't be pulled into a common method without adding reflection.

Reply:
:(